[FLINK-19546][doc] Add documentation for native Kubernetes HA (#14006)
apache/flink: docs/ops/jobmanager_high_availability.md

{% highlight bash %}
$ ./bin/flink run-application -p 8 -t kubernetes-application \
  -Dkubernetes.cluster-id=<ClusterId> \
  -Dtaskmanager.memory.process.size=4096m \
  -Dkubernetes.taskmanager.cpu=2 \
  -Dtaskmanager.numberOfTaskSlots=4 \
  -Dkubernetes.container.image=<CustomImageName> \
  -Dhigh-availability=org.apache.flink.kubernetes.highavailability.KubernetesHaServicesFactory \
  -Dhigh-availability.storageDir=s3://flink/flink-ha \
  -Drestart-strategy=fixed-delay -Drestart-strategy.fixed-delay.attempts=10 \
  -Dcontainerized.master.env.ENABLE_BUILT_IN_PLUGINS=flink-s3-fs-hadoop-{{site.version}}.jar \
  -Dcontainerized.taskmanager.env.ENABLE_BUILT_IN_PLUGINS=flink-s3-fs-hadoop-{{site.version}}.jar \
  local:///opt/flink/examples/streaming/StateMachineExample.jar
{% endhighlight %}

### High availability data clean up
Currently, when a Flink cluster reached the terminal state(`FAILED`, `CANCELED`, `FINISHED`), all the HA data, including metadata in Kubernetes ConfigMap and HA storage on DFS, will be cleaned up. So if the user stops the Flink session/application cluster manually(via the following two commands), the Flink cluster could not be recovered.

{% highlight bash %}

> $ echo 'stop' | ./bin/kubernetes-session.sh -Dkubernetes.cluster-id=<ClusterId> -Dexecution.attached=true

Author: Hmm. After checking the implementation, the above command will only shut down the cluster and not clean up the HA state. And we are not deleting the HA related ConfigMap here.

Only when the job reached to the terminal state, the HA data will be cleaned up.

I will update the documentation.

   <pre>
$ bin/yarn-session.sh -n 2</pre>

## Kubernetes Cluster High Availability

> We are using a [deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) for the JobManager when deploying Flink on Kubernetes cluster. The `replicas` is configured to 1, which means that a new JobManager will be launched to take over the leadership once the current one terminated exceptionally.

Reviewer: I'm a little confused by this sentence, who is "we" referring to? Does this make sense? 

```suggestion
When running Flink as a Kubernetes [deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/), the `replica` count should be configured to 1.
This means that a new JobManager will be launched to take over leadership if the current one terminates exceptionally. 
```

Author: Hmm, I think "running Flink as a Kubernetes deployment" is not accurate. Because we are running Flink JobManager as a Kubernetes deployment. Do you think the following sentences make sense?

When running Flink JobManager as a Kubernetes [deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/), the `replica` count should be configured to 1 or greater.
* The value `1` means that a new JobManager will be launched to take over leadership if the current one terminates exceptionally. 
* The value `N`(greater than 1) means that multiple JobManagers will be launched while one is active and others are standby. Starting more than one JobManager will make the recovery faster.

### Configuration
{% highlight yaml %}
kubernetes.cluster-id: <ClusterId>
high-availability: org.apache.flink.kubernetes.highavailability.KubernetesHaServicesFactory
high-availability.storageDir: hdfs:///flink/recovery
{% endhighlight %}

#### Example: Highly Available Standalone Flink Cluster on Kubernetes

> Both session and job/application clusters could use the Kubernetes high availability service. Users just need to add the following Flink config options to [flink-configuration-configmap.yaml]({{ site.baseurl}}/ops/deployment/kubernetes.html#common-cluster-resource-definitions). All other yamls do not need to be updated.

Reviewer: ```suggestion
Both session and job/application clusters support using the Kubernetes high availability service. Users just need to add the following Flink config options to [flink-configuration-configmap.yaml]({{ site.baseurl}}/ops/deployment/kubernetes.html#common-cluster-resource-definitions). All other yamls do not need to be updated.
```

> <span class="label label-info">Note</span> You need to have hadoop in the classpath or enable the corresponding FileSystem plugin for the Flink image. Refer to [custom Flink image]({{ site.baseurl}}/ops/deployment/docker.html#customize-flink-image) and [enable plugins]({{ site.baseurl}}/ops/deployment/docker.html#using-plugins) for more information.

Reviewer: This shouldn't be Hadoop specific. 
```suggestion
<span class="label label-info">Note</span> The filesystem which corresponds to the scheme of your configured ha storage directory must be available to the runtime. Refer to [custom Flink image]({{ site.baseurl}}/ops/deployment/docker.html#customize-flink-image) and [enable plugins]({{ site.baseurl}}/ops/deployment/docker.html#using-plugins) for more information.
```

  -Dcontainerized.master.env.ENABLE_BUILT_IN_PLUGINS=flink-s3-fs-hadoop-{{site.version}}.jar \
  -Dcontainerized.taskmanager.env.ENABLE_BUILT_IN_PLUGINS=flink-s3-fs-hadoop-{{site.version}}.jar \
  local:///opt/flink/examples/streaming/StateMachineExample.jar
{% endhighlight %}

### High Availability Data Clean Up
Currently, when a Flink job reached the terminal state (`FAILED`, `CANCELED`, `FINISHED`), all the HA data, including metadata in Kubernetes ConfigMap and HA state on DFS, will be cleaned up.

So the following command will only shut down the Flink session cluster and leave all the HA related ConfigMaps, state untouched.
{% highlight bash %}
$ echo 'stop' | ./bin/kubernetes-session.sh -Dkubernetes.cluster-id=<ClusterId> -Dexecution.attached=true
{% endhighlight %}

The following commands will cancel the job in application or session cluster and effectively remove all its HA data.
{% highlight bash %}
# Cancel a Flink job in the existing session
$ ./bin/flink cancel -t kubernetes-session -Dkubernetes.cluster-id=<ClusterID> <JobID>
# Cancel a Flink application
$ ./bin/flink cancel -t kubernetes-application -Dkubernetes.cluster-id=<ClusterID> <JobID>
{% endhighlight %}

> If the user wants to keep the HA data and restart the Flink cluster, he/she could also simply delete the deploy(via `kubectl delete deploy <ClusterID>`). All the Flink cluster related resources will be destroyed(e.g. JobManager Deployment, TaskManager pods, services, Flink conf ConfigMap) so that it will not occupy the Kubernetes cluster resources. For the HA related ConfigMaps, we do not set the owner reference and they could be retained. Then he/she could use `kubernetes-session.sh` or `flink run-application` to start the session/application again. All the previous suspending running jobs could recover from the latest checkpoint successfully.

Reviewer: A good rule of thumb, if you ever find yourself writing "he/she" then just use "they" instead.

Also, the word "could" implies that something only works some of the time but I don't think that's what you mean to say here. 

```suggestion
To keep HA data while restarting the Flink cluster,  simply delete the deploy (via `kubectl delete deploy <ClusterID>`). All the Flink cluster related resources will be destroyed (e.g. JobManager Deployment, TaskManager pods, services, Flink conf ConfigMap), as to not occupy the Kubernetes cluster resources. However, HA related ConfigMaps do not set the owner reference and they will be retained. When restarting the session / application, use `kubernetes-session.sh` or `flink run-application`. All the previous suspending running jobs will recover from the latest checkpoint successfully.
```

  -Dcontainerized.master.env.ENABLE_BUILT_IN_PLUGINS=flink-s3-fs-hadoop-{{site.version}}.jar \
  -Dcontainerized.taskmanager.env.ENABLE_BUILT_IN_PLUGINS=flink-s3-fs-hadoop-{{site.version}}.jar \
  local:///opt/flink/examples/streaming/StateMachineExample.jar
{% endhighlight %}

### High Availability Data Clean Up
Currently, when a Flink job reached the terminal state (`FAILED`, `CANCELED`, `FINISHED`), all the HA data, including metadata in Kubernetes ConfigMap and HA state on DFS, will be cleaned up.

So the following command will only shut down the Flink session cluster and leave all the HA related ConfigMaps, state untouched.
{% highlight bash %}
$ echo 'stop' | ./bin/kubernetes-session.sh -Dkubernetes.cluster-id=<ClusterId> -Dexecution.attached=true
{% endhighlight %}

The following commands will cancel the job in application or session cluster and effectively remove all its HA data.
{% highlight bash %}
# Cancel a Flink job in the existing session
$ ./bin/flink cancel -t kubernetes-session -Dkubernetes.cluster-id=<ClusterID> <JobID>
# Cancel a Flink application
$ ./bin/flink cancel -t kubernetes-application -Dkubernetes.cluster-id=<ClusterID> <JobID>
{% endhighlight %}

> To keep HA data while restarting the Flink cluster, simply delete the deploy (via `kubectl delete deploy <ClusterID>`). All the Flink cluster related resources will be destroyed (e.g. JobManager Deployment, TaskManager pods, services, Flink conf ConfigMap), as to not occupy the Kubernetes cluster resources. However, HA related ConfigMaps do not set the owner reference and they will be retained. When restarting the session / application, use `kubernetes-session.sh` or `flink run-application`. All the previous suspending running jobs will recover from the latest checkpoint successfully.

Reviewer 2: maybe: deploy -> deployment in the text?

> To keep HA data while restarting the Flink cluster, simply delete the deploy (via `kubectl delete deploy <ClusterID>`). All the Flink cluster related resources will be destroyed (e.g. JobManager Deployment, TaskManager pods, services, Flink conf ConfigMap), as to not occupy the Kubernetes cluster resources. However, HA related ConfigMaps do not set the owner reference and they will be retained. When restarting the session / application, use `kubernetes-session.sh` or `flink run-application`. All the previous suspending running jobs will recover from the latest checkpoint successfully.

Reviewer 2: maybe: will be deleted

Reviewer 2: For future documentation work: It is simpler to review if you put every sentence in its own line. That way it is much easier to see where the comment belongs to.

Reviewer 2: I don't understand the subclause `, as to not occupy .... resources`.

Reviewer 2: HA related ConfigMaps will be retained because they do not set the owner reference.

Reviewer 2: When restarting the session / application using `kubernetes-session.sh` or `flink run-application`, all previously running jobs will be recovered and restart from the latest successful checkpoint.

Author: > I don't understand the subclause , as to not occupy .... resources.

What I want to state here is that we will free all the allocated resources from K8s so that they could be used by others. At the same time, we still have the ability to recover the Flink application at any time if you want.

Reviewer 2: I think we don't need the `as to not occupy ... resources` subclause. Are you also addressing my other comments for this paragraph?

Author: Yeah. I have addressed the comments and force pushed.

> To keep HA data while restarting the Flink cluster, simply delete the deploy (via `kubectl delete deploy <ClusterID>`). All the Flink cluster related resources will be destroyed (e.g. JobManager Deployment, TaskManager pods, services, Flink conf ConfigMap), as to not occupy the Kubernetes cluster resources. However, HA related ConfigMaps do not set the owner reference and they will be retained. When restarting the session / application, use `kubernetes-session.sh` or `flink run-application`. All the previous suspending running jobs will recover from the latest checkpoint successfully.

Reviewer 2: ```suggestion
To keep HA data while restarting the Flink cluster, simply delete the deploy (via `kubectl delete deploy <ClusterID>`). All the Flink cluster related resources will be deleted (e.g. JobManager Deployment, TaskManager pods, services, Flink conf ConfigMap). HA related ConfigMaps will be retained because they do not set the owner reference. When restarting the session / application using `kubernetes-session.sh` or `flink run-application`, all previously running jobs will be recovered and restarted from the latest successful checkpoint.
```
